pyads/symbol.py

```diff
@@ -0,0 +1,271 @@
+    :ivar comment: Comment of symbol
+    """
+
+    def __init__(
+        self,
+        plc: "Connection",
+        name: Optional[str] = None,
+        index_group: Optional[int] = None,
+        index_offset: Optional[int] = None,
+        symbol_type: Optional[str] = None,
+        comment: Optional[str] = None,
+    ) -> None:
+        """Create AdsSymbol instance
+
+        Specify either the variable name or the index_group **and**
+        index_offset so the symbol can be located.
+        If the name was specified but not all other attributes were,
+        the other attributes will be looked up from the connection.
+        `symbol_type` should be a string representing a PLC type (e.g.
+        'LREAL').
+
+        :param plc: Connection instance
+        :param name:
+        :param index_group:
+        :param index_offset:
+        :param symbol_type: PLC variable type (e.g. 'LREAL')
+        :param comment:
+        """
+        self._plc = plc
+
+        self._handles_list: List[Tuple[int, int]] = []  # Notification handles
+
+        do_lookup = True
+
+        if index_group is None or index_offset is None or symbol_type is None:
+            if name is None:
+                raise ValueError(
+                    "Please specify either `name`, or "
+                    "`index_group`, `index_offset` and "
+                    "plc_type"
+                )
+        else:
+            # We have an address and the type, so we don't need to do a lookup
+            do_lookup = False
+
+        self.name = name
+        self.index_offset = index_offset
+        self.index_group = index_group
+        self.symbol_type = symbol_type
+        self.comment = comment
+
+        if do_lookup:
+            self.create_symbol_from_info()  # Perform remote lookup
+
+        # Now `self.symbol_type` should have a value, find the actual PLCTYPE
+        # from it.
+        # This is relevant for both lookup and full user definition.
+
+        self.plc_type: Optional[Any] = None
+        if self.symbol_type is not None:
+            self.plc_type = self.get_type_from_str(self.symbol_type)
+
+    def create_symbol_from_info(self) -> None:
+        """Look up remaining info from the remote
+
+        The name must already be present.
+        Other values will already have a default value (mostly None).
+        """
+        info = adsGetSymbolInfo(self._plc._port, self._plc._adr, self.name)
+
+        self.index_group = info.iGroup
+        self.index_offset = info.iOffs
+        if info.comment:
+            self.comment = info.comment
+
+        # info.dataType is an integer mapping to a type in
+        # constants.ads_type_to_ctype.
+        # However, this type ignores whether the variable is really an array!
+        # So are not going to be using this and instead rely on the textual
+        # type
+        self.symbol_type = info.symbol_type  # Save the type as string
+
+    def read_write_check(self):
```

Comment:
We can omit the read_write_check? We should do all checks on initialization. So I see no obvious reason to check again especially as the check is done on every read/write.

Reply:
You have a point. I figured that, because the connection will be owned outside the Symbol object, this could result into confusing problems. For example:

```
plc = ...

with plc:
    # Connection is now opened
    symbol = plc.get_symbol(...)

    # ... lot's of code

# User tries to use the symbol again, but connection has been closed now:
symbol.read()
```

Do we want to catch something like this?

Reply:
Just checked this by the way. WIthout the `read_write_check`, the Connection quietly returns `None` when it has been closed. This might lead to sneaky bugs in users code.

Thoughts?

Reply:
Hm, I actually prefer if Connection throws the exception but this breaks compatibility and might cause issues in some use-cases. So maybe we just check for plc.open on read/write and throw an exception if it is not open. What do you think)

Reply:
I agree. So we check `plc.is_open` on every read and write then?

Maybe we can add an `__init__` keyword argument to `Connection` for compatibility. Something like `read_write_error=False` that determines whether errors are thrown.

Reply:
> Maybe we can add an __init__ keyword argument to Connection for compatibility. Something like read_write_error=False that determines whether errors are thrown.

That would be an option but as Connection is the parent of Symbol there is no way to force that a Connection object is created with this flag set. I think we get the least trade-off if we just check for plc.is_open on each read/write. This is not much computation anyway.